packages/@lwc/template-compiler/src/parser/parser.ts

```diff
@@ -134,33 +158,147 @@ export default class ParserCtx {
     }
 
-    private currentScope(): ParentNode[] | undefined {
-        return this.scopes[this.scopes.length - 1];
+    hasSeenSlot(name: string): boolean {
+        return this.hasSeenSlotInAncestorIfTree(name);
+    }
+
+    addSeenSlot(name: string): void {
+        const currentSeenSlots = this.seenSlotsFromAncestorIfTree();
+        if (currentSeenSlots) {
+            currentSeenSlots.add(name);
+        } else {
+            this.seenSlots.add(name);
+        }
+    }
+
+    private currentElementScope(): ParentNode[] | undefined {
+        return this.elementScopes[this.elementScopes.length - 1];
+    }
+
+    beginSiblingScope() {
+        this.siblingScopes.push({
+            ancestorIfContext: this.currentIfContext() || this.ancestorIfContext(),
+        });
+    }
+
+    endSiblingScope() {
+        this.siblingScopes.pop();
+    }
+
+    // Next goal is to try and move some of the "business logic" from here out into the parser's index.ts itself
+    // so that ParserCtx can be more of an utility than handling specific logic for how to handle the nodes.
+    beginIfChain(node: IfBlock) {
+        const currentSiblingContext = this.currentSiblingContext();
+        if (!currentSiblingContext) {
+            throw new Error('Cannot invoke beginIfChain if there is currently no sibling context');
+        }
+
+        // An if block always starts a new chain.
+        if (this.isParsingIfBlock()) {
+            this.endIfChain();
+        }
+
+        const previouslySeenSlots = this.seenSlotsFromAncestorIfTree();
+        currentSiblingContext.ifContext = {
+            currentNode: node,
+            seenSlots: [
+                previouslySeenSlots ? new Set<string>(previouslySeenSlots) : new Set<string>(),
+            ],
+        };
+    }
+
+    updateIfChain(node: ElseifBlock | ElseBlock) {
+        const currentIfContext = this.currentIfContext();
+        if (!currentIfContext) {
+            throw new Error();
+        }
+
+        currentIfContext.currentNode = node;
+
+        const previouslySeenSlots = this.seenSlotsFromAncestorIfTree();
+        currentIfContext.seenSlots.push(
+            previouslySeenSlots ? new Set<string>(previouslySeenSlots) : new Set<string>()
+        );
+    }
+
+    endIfChain() {
+        const currentIfContext = this.currentIfContext();
+        if (!currentIfContext) {
+            return;
+        }
+
+        // Merge seen slot names from the current if chain into the parent scope.
+        const seenSlotsInAncestorIfTree = this.seenSlotsFromAncestorIfTree();
+        for (const seenSlots of currentIfContext.seenSlots) {
+            for (const name of seenSlots) {
+                seenSlotsInAncestorIfTree.add(name);
+            }
+        }
+
+        const currentSiblingContext = this.currentSiblingContext();
+        if (currentSiblingContext) {
+            currentSiblingContext.ifContext = undefined;
+        }
+    }
+
+    getSiblingIfNode(): ParentNode | undefined {
```

Comment:
I will be a little more restrictive about the types here: 

```suggestion
    getSiblingIfNode(): IfBlock | ElseifBlock | ElseBlock | undefined {
```